File: src/Go/Prime.go
package main

import (
	"fmt"
	"math"
)

var isDebug = false

type Prime struct {
	_maxInd   uint64
	_maxKeep  int
	seqList   *[]string
	interList *[]string
	prevNo    uint64
	_prime    *[]uint64
	_offSet   uint64
}

func newPrime(page uint64, repeat uint64) *Prime {
	var p Prime
	p._maxKeep = int(math.Sqrt(float64(page*repeat)) / math.Log(math.Sqrt(float64(page*repeat))) * 1.3)
	sp := make([]uint64, p._maxKeep+800000)
	p._prime = &sp
	p.interList = &[]string{}
	p.seqList = &[]string{}
	return &p
}

func (p Prime) get(index int) uint64 {
	return (*p._prime)[index]
}

func (p *Prime) add(n uint64) {
	(*p._prime)[(*p)._maxInd-(*p)._offSet] = n
	(*p)._maxInd++
}

func (p *Prime) size() int {
	return int((*p)._maxInd)
}

func (p *Prime) generateResults(inter uint64, endNo uint64) {
	(*p).outputSequence(p.prevNo, endNo)
	(*p).outputInterval(inter)
	(*p).prevNo = endNo
	(*p).freeUp()
}

func (p Prime) outputInterval(inter uint64) {
	var s string
	if inter%uint64(math.Pow10(len(fmt.Sprintf("%d", inter))-1)) == 0 {
		s = dfString(inter) + "|" + fmt.Sprintf("%d", p._maxInd) + "|" + fmt.Sprintf("%d", (*p._prime)[p._maxInd-p._offSet-1])
		*p.interList = append(*p.interList, s)
		if isDebug { fmt.Println("[In:]", s) }
	}
}

func (p Prime) outputSequence(beginNo uint64, endNo uint64) {
	var s string
	for i := len(fmt.Sprintf("%d", beginNo)) - 1; i <= len(fmt.Sprintf("%d", endNo))-1; i++ {
		for j := 1; j < 10; j++ {
			seq := uint64(j) * uint64(math.Pow10(i))
			if seq < beginNo { continue	}
			if seq >= endNo { break	}
			v := (*p._prime)[p._maxInd-p._offSet-1-(endNo-seq)]
			s = dfString(seq) + "|" + fmt.Sprintf("%d", v)
			*p.seqList = append(*p.seqList, s)
			if isDebug { fmt.Println("==>[No:]", s)	}
		}
	}
}

func (p *Prime) freeUp() {
	if len(*p._prime) > (*p)._maxKeep {
		(*p)._offSet = (*p)._maxInd - uint64((*p)._maxKeep)
	}
}

func (p Prime) printTable() {
	fmt.Println("## 素数区间表")
	fmt.Println("区间|个数|最大值")
	fmt.Println("---|---|---")
	for _, s := range *p.interList { fmt.Println(s)	}
	fmt.Println("## 素数序列表")
	fmt.Println("序号|数值")
	fmt.Println("---|---")
	for _, s := range *p.seqList { fmt.Println(s) }
}

func dfString(l uint64) string {
	s := fmt.Sprintf("%d", l)
	if l%(10000*10000*10000) == 0 {
		s = s[:len(s)-12] + "万亿"
	} else if l%(10000*10000) == 0 {
		s = s[:len(s)-8] + "亿"
	} else if l%10000 == 0 {
		s = s[:len(s)-4] + "万"
	}
	return s
}

File: src/Ruby/prime.rb
class Prime

  def initialize (limit)
    @is_debug = false
    @max_keep = (Math.sqrt(limit) / Math.log(Math.sqrt(limit)) * 2.5).to_i
    @prime_list = []
    @max_ind = 0
    @prev_no = 0
    @seq_list = []
    @inter_list = []
  end

  def get(index)
    @prime_list[index]
  end

  def size
    @max_ind
  end

  def add(p)
    @prime_list.push(p)
    @max_ind = @max_ind + 1
  end

  def output_interval(inter)
    if inter % 10 ** (inter.to_s.length - 1) == 0
      s = df_string(inter) + "|" + @max_ind.to_s + "|" + @prime_list[@prime_list.length - 1].to_s
      @inter_list.push(s)
      puts "[In:]" + s if @is_debug
    end
  end

  def output_sequence (beginNo, endNo)
    for i in beginNo.to_s.length - 1..endNo.to_s.length - 1
      for j in 1..9
        seq = j * (10 ** i)
        next if seq < beginNo
        return if seq >= endNo
        l = @prime_list[@prime_list.length - 1 - (endNo - seq)]
        s = df_string(seq) + "|" + l.to_s
        @seq_list.push(s)
        puts "==>[No:]" + s if @is_debug
      end
    end
  end

  def free_up
    if @max_ind > @max_keep
      @prime_list = @prime_list[0, @max_keep - 1]
    end
  end

  def generate_results(inter, endNo)
    output_sequence(@prev_no, endNo)
    output_interval(inter)
    @prev_no = endNo
    free_up
  end

  def df_string(l)
    s = l.to_s
    if l % 1_0000_0000_0000 == 0
      s = s[0, s.length - 12] + "万亿"
    elsif l % 1_0000_0000 == 0
      s = s[0, s.length - 8] + "亿"
    elsif l % 1_0000 == 0
      s = s[0, s.length - 4] + "万"
    end
    s
  end

  def print_table
    puts "## 素数区间表"
    puts "区间|个数|最大值"
    puts "---|---|---"
    @inter_list.each { |s| puts s }
    puts "## 素数序列表"
    puts "序号|数值"
    puts "---|---"
    @seq_list.each { |s| puts s }
  end

end

